### PR TITLE
Add Check for too-large Bond Supply

### PR DESCRIPTION
**Pull request by Namaskar-1F64F**
https://github.com/spearbit-audits/porter/issues/17

Description:

Issuing too many bonds can result in users being unable to redeem. This is caused by arithmetic overflow in previewRedeemAtMaturity.

If a user's bonds and the paidAmount's (or bonds * nonPaidAmount) product is greater than 2**256, it will overflow, reverting all attempts to redeem bonds.

This is fixed by adding the safety checks

### CODE REVIEW

File: test/constants.ts [hunk 1]
```diff
@@ -29,7 +29,9 @@ export const ELEVEN_YEARS_FROM_NOW_IN_SECONDS = Math.round(
 
 export const ONE = utils.parseUnits("1", 18);
 export const WAD = utils.parseUnits("1", 18);
-
+export const SQRT_MAX_UINT256 = BigNumber.from(
+  "340282366920938463463374607431768211456"
```
Comment (RusseII) [inline test/constants.ts:33]: Nice. I ran through our test suite as well with SQRT_MAX_UINT256.sub(1) as maxSupply for bond and nothing explicitly broke (except for rounding errors)

**Review by RusseII — APPROVED** [review]
Looks good. In addition to this we may want to add some tests for additional overflow situations in another PR. 

We should make sure it's not possible to overflow in 
```
collateralTokensToSend = collateralRatio.mulWadDown(
            bonds.mulDivDown(nonPaidAmount, bondSupply)
        );
```

File: contracts/Bond.sol [hunk 2]
```diff
@@ -94,6 +94,13 @@ contract Bond is
         uint256 _convertibleRatio,
         uint256 maxSupply
     ) external initializer {
+        /*
+            Safety check: Ensure multiplication can not overflow uint256.
+        */
+        maxSupply * maxSupply;
+        maxSupply * _collateralRatio;
+        maxSupply * _convertibleRatio;
```
Comment (RusseII) [inline contracts/Bond.sol:100]: The `maxSupply * _convertibleRatio;` is probably unnecessary since we have another check that  _convertibleRatio <= _collateralRatio in bondFactory... 

Gas cost is negligible.... let's leave it here for now to make sure all the multiplication cases are covered. In the future let's explore moving all the bond invariants into the initializer function instead of `createBond`
